File: Visual Studio Product Team/Official Visual Studio 2008 VB.NET Samples/[VB]-Official Visual Studio 2008 VB.NET Samples/Expression Tree Visualizer Sample (Visual Basic)/VB/ExpressionTreeVisualizer/ExpressionTreeVisualizer.vb
Imports System.Linq.Expressions
imports Microsoft.VisualStudio.DebuggerVisualizers

<Assembly: DebuggerVisualizer(GetType(ExpressionTreeVisualizer), GetType(ExpressionTreeVisualizerObjectSource), Target:=GetType(Expression), Description:="Expression Tree Visualizer")> 

Public Class ExpressionTreeVisualizer
    Inherits DialogDebuggerVisualizer

    Private m_modalService As IDialogVisualizerService

    Protected Overrides Sub Show(ByVal windowService As IDialogVisualizerService, ByVal objectProvider As IVisualizerObjectProvider)
        m_modalService = windowService
        If m_modalService Is Nothing Then
            Throw New ApplicationException("This debugger does not support modal visualizers")
        End If

        Dim container As ExprTreeContainer = objectProvider.GetObject()
        Dim treeForm = New ExpressionTreeForm(container, container.Expression)
        m_modalService.ShowDialog(treeForm)
    End Sub

End Class

File: Visual Studio Product Team/Official Visual Studio 2008 C# Samples/[C#]-Official Visual Studio 2008 C# Samples/ExpressionTreeVisualizer/C#/ExpressionTreeVisualizer/ExpressionTreeVisualizer.cs
using System;
using System.Text;
using System.Drawing;
using System.IO;
using System.Reflection;
using System.Windows.Forms;
using Microsoft.CSharp;
using System.Collections;
using System.Linq.Expressions;
using Microsoft.VisualStudio.DebuggerVisualizers;
using System.Diagnostics;
using System.Runtime.Serialization;
using ExpressionVisualizer;

[assembly: DebuggerVisualizer(typeof(ExpressionTreeVisualizer), typeof(ExpressionTreeVisualizerObjectSource), Target = typeof(Expression), Description = "Expression Tree Visualizer")]

namespace ExpressionVisualizer
{

    public class ExpressionTreeVisualizer : DialogDebuggerVisualizer
    {
        private IDialogVisualizerService modalService;

        protected override void Show(IDialogVisualizerService windowService, IVisualizerObjectProvider objectProvider)
        {

            modalService = windowService;
            if (modalService == null)
                throw new NotSupportedException("This debugger does not support modal visualizers");

            ExpressionTreeContainer container = (ExpressionTreeContainer)objectProvider.GetObject();
            TreeBrowser browser = new TreeBrowser();
            browser.Add(container.Tree);
            TreeWindow treeForm = new TreeWindow(browser, container.Expression);
            modalService.ShowDialog(treeForm);
        }
    }
}